dat/CMakeLists.txt

```diff
@@ -0,0 +1,143 @@
+set(VARDATND)
+
+set(VARDATD
+    bogusmon
+    data
+    engrave
+    epitaph
+    oracles
+    options
+    quest.lua
+    rumors)
+
+set(VARDAT ${VARDATD} ${VARDATND})
+
+set(DATHELP
+    help
+    hh
+    cmdhelp
+    keyhelp
+    history
+    opthelp
+    optmenu
+    wizhelp)
+
+file(
+  GLOB SPEC_LEVS
+  RELATIVE ${NH_DAT}
+  asmodeus.lua
+  baalz.lua
+  bigrm-*.lua
+  castle.lua
+  fakewiz?.lua
+  juiblex.lua
+  knox.lua
+  medusa-?.lua
+  minend-?.lua
+  minefill.lua
+  minetn-?.lua
+  oracle.lua
+  orcus.lua
+  sanctum.lua
+  soko?-?.lua
+  tower?.lua
+  valley.lua
+  wizard?.lua
+  nhcore.lua
+  nhlib.lua
+  themerms.lua
+  astral.lua
+  air.lua
+  earth.lua
+  fire.lua
+  water.lua)
+
+file(
+  GLOB QUEST_LEVS
+  RELATIVE ${NH_DAT}
+  ???-goal.lua ???-fil?.lua ???-loca.lua ???-strt.lua)
+
+set(DATNODLB ${VARDATND} license symbols)
+
+set(DATDLB ${DATHELP} dungeon.lua tribute ${SPEC_LEVS} ${QUEST_LEVS} ${VARDATD})
+
+set(DAT ${DATNODLB} ${DATDLB})
+
+file(MAKE_DIRECTORY ${NH_DAT_GEN})
+
+# Inputs to makedef.
+set(ALL_DAT_NOTGEN
+    ${DATHELP}
+    ${SPEC_LEVS}
+    ${QUEST_LEVS}
+    bogusmon.txt
+    data.base
+    dungeon.lua
+    engrave.txt
+    epitaph.txt
+    oracles.txt
+    quest.lua
+    rumors.fal
+    rumors.tru
+    tribute)
+
+# makedef and friends like having their inputs in the same directory.
+foreach(filename ${ALL_DAT_NOTGEN})
+  configure_file(${NH_DAT}/${filename} ${filename} COPYONLY)
+endforeach(filename)
+
+# Dat files that are not inputs to any tools.
+set(ALL_DAT_NOTGEN ${ALL_DAT_NOTGEN} ${DATNODLB})
+
+add_custom_command(
+  DEPENDS $<TARGET_FILE:makedefs> bogusmon.txt engrave.txt epitaph.txt
+  OUTPUT bogusmon epitaph engrave
+  COMMAND $<TARGET_FILE:makedefs> ARGS -s)
```

Comment:
Can you clarify this comment? Are you suggesting a change like this?

```diff
-   COMMAND $<TARGET_FILE:makedefs> ARGS -s)
+   COMMAND makedefs ARGS -s)
```

I wasn't sure that this would include the full path. But perhaps I am misunderstanding your suggestion.

Reply:
Yes, that's what I meant. This is a documented feature:

>If ``COMMAND`` specifies an executable target name (created by the ``add_executable()`` command), it will automatically be replaced by the location of the executable created at build time if .. [cross compile details] ..

This is how I use `add_custom_command` most of the time. Alternatively I put a variable that contains the path to a program that was found via `find_program`.

The target name syntax could be confused with referring to a program that is on the `$PATH`. But that's not a common use case in my experience.

I should clarify that this is just a minor nitpick. I've never used TARGET_FILE for this and don't remember seeing it anywhere else either, but it's readable and I'm sure it does the job. :wink:

Reply:
Interesting! But the cmake documentation continues [like this](https://cmake.org/cmake/help/latest/command/add_custom_command.html#:~:text=If%20COMMAND%20specifies%20an%20executable,is%20not%20set%20to%20true):

> Arguments to `COMMAND` may use generator expressions. Use the `TARGET_FILE` generator expression to refer to the location of a target later in the command line (i.e. as a command argument rather than as the command to execute).

(OK, so in our case it is the command itself, so I guess we shouldn't use `TARGET_FILE`?)

> Whenever one of the following target based generator expressions are used as a command to execute or is mentioned in a command argument, a target-level dependency will be added automatically so that the mentioned target will be built before any target using this custom command (see policy CMP0112).
> 
> ```
> TARGET_FILE
> TARGET_LINKER_FILE
> TARGET_SONAME_FILE
> TARGET_PDB_FILE
> ```

So *this* style introduces a dependency (which we want in this case).

Overall, cmake is such a highly non-uniform language, it's kind of sad it became standard (over, say, something closer to bazel).

Reply:
Interesting, I hadn't seen that but some testing reveals that it doesn't quite work how I'd expect.

I think I get it now.. For the `COMMAND` argument, using either the target name or `$<TARGET_FILE:target_name>` creates a *target-level* dependency but neither introduce a *file-level* dependency.

>  This target-level dependency does NOT add a file-level dependency that would cause the custom command to re-run whenever the executable is recompiled. List target names with the ``DEPENDS`` option to add such file-level dependencies.

File-level dependence is what I had in mind in my previous comments. TBH I didn't know CMake had different kinds of dependencies.

I like CMake but it definitely has its sharp edges (and mushy ones.).

Reply:
I am not sure I understand the difference between target-level dependencies and file-level dependencies, but I gather the latter means: If a target has a file-level dependency on a binary (file), it will cause that binary file to be build, but if it's already present it is not rebuild?

By the way if memory serves some of the `$<TARGET_FILE:target_name>` syntax I [introduced](https://github.com/heiner/NetHack/commit/226be9db5a45eec992c4f4c1e70e589234abb8bf) after I noticed my cmake setup breaks when applied to a NetHack which also has build artefacts in-source (e.g., run make, then run cmake). CMake chose the wrong files to depend on in some cases, [this commit](https://github.com/heiner/NetHack/commit/226be9db5a45eec992c4f4c1e70e589234abb8bf) tried to work around that, but might have overdone it.

Reply:
Yes from what I can tell that is the difference between target and file-level dependencies. Rebuilding the data files when the binary has change makes most sense to me.

Hm, that's odd. I may play around with the scenario you described. Was the CMake build an in-source or out-of-source build? Anyhow, keeping it as `TARGET_FILE` is good. Please consider this conversation resolved

Reply:
Ah I'm sorry, I wasn't quite clear.

Issues can occur when calling cmake after using NetHacks current/standard Makefile setup, e.g., on Linux:

```sh
cd sys/unix
sh setup.sh hints/linux
cd ../..
make
mkdir build && cd build
cmake .. -GNinja
ninja
```

The issue is that this causes in-source build artefacts to be picked up by cmake, as opposed to their out-of-source targets. The commit I linked above is trying to work around that; I don't remember if using `$<TARGET_FILE:target_name>` truly helped with this; what definitely helped was using absolute paths everywhere.